org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/InlayHintVisitor.java

```diff
@@ -0,0 +1,261 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Microsoft Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.handlers;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import com.google.common.base.Objects;
+
+import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.ASTVisitor;
+import org.eclipse.jdt.core.dom.BooleanLiteral;
+import org.eclipse.jdt.core.dom.CastExpression;
+import org.eclipse.jdt.core.dom.CharacterLiteral;
+import org.eclipse.jdt.core.dom.ClassInstanceCreation;
+import org.eclipse.jdt.core.dom.EnumConstantDeclaration;
+import org.eclipse.jdt.core.dom.Expression;
+import org.eclipse.jdt.core.dom.IMethodBinding;
+import org.eclipse.jdt.core.dom.IVariableBinding;
+import org.eclipse.jdt.core.dom.MethodInvocation;
+import org.eclipse.jdt.core.dom.NullLiteral;
+import org.eclipse.jdt.core.dom.NumberLiteral;
+import org.eclipse.jdt.core.dom.SimpleName;
+import org.eclipse.jdt.core.dom.StringLiteral;
+import org.eclipse.jdt.core.dom.TypeLiteral;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import org.eclipse.lsp4j.proposed.InlayHint;
+
+@SuppressWarnings("unchecked")
+class InlayHintVisitor extends ASTVisitor {
+	private List<InlayHint> hints;
+	private int startOffset;
+	private int endOffset;
+	private ITypeRoot typeRoot;
+	private PreferenceManager preferenceManager;
+
+	InlayHintVisitor(int startOffset, int endOffset, ITypeRoot typeRoot, PreferenceManager preferenceManager) {
+		this.startOffset = startOffset;
+		this.endOffset = endOffset;
+		this.typeRoot = typeRoot;
+		this.hints = new ArrayList<>();
+		this.preferenceManager = preferenceManager;
+	}
+
+	@Override
+	public boolean visit(EnumConstantDeclaration node) {
+		if (isOutOfRange(node)) {
+			return true;
+		}
+
+		resolveInlayHints(node.resolveConstructorBinding(), node.arguments());
+
+		return true;
+	}
+
+	@Override
+	public boolean visit(ClassInstanceCreation node) {
+		if (isOutOfRange(node)) {
+			return true;
+		}
+
+		resolveInlayHints(node.resolveConstructorBinding(), node.arguments());
```

Comment:
The lsp specification provide a resolveInlayHint.I wonder if it should be better to resolve binding in this lsp method?

Reply:
Rename to `getInlayHintFromMethodBinding()` ?

Reply:
No I mean that your current visiter create an inlayhint with à given position but without label and label is resolved in inlayHint/resolve https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHint_resolve

Reply:
It might not be a good idea to put the label calculation into the `resolve` request.

The label calculation for each node is relatively 'cheap'. And by calculating the label in the `textDocument/inlayHint` request route, we only need to traverse the AST tree once.

If we put the label to the resolve request route. we have to perform `N` times traverse to the AST tree to find the node, which might cause perf downgrade.

It's also not suggested to do that in LSP spec:

_The request is sent from the client to the server to resolve additional information for a given inlay hint. This is usually used to compute the **tooltip**, **location** or **command** properties_

This behavior is commonly used in other LSP requests/responses. For example, LS can calculate labels first for each completion items. And resolve the text edit/command in the resolve requests.

We can consider to leverage the resolve capability in the future if we want to expand the functionality of the inlay hints.

Reply:
Thank a lot @jdneo for your great feedback. I would like just to highlight this resolve capability but your comment is relevant and I agree with you.

Reply:
Thanks for the discussion too :)

I'm also interested what kind of things are worth to be put into `tooltip` and `command` of the inlay hints. Please let me know if you have any ideas on that in the future.

Reply:
I think the main idea is to display hint as soon as possible. Imagine that you have 10 hints to display . 9 hints are Quick to resolve and 1 hint is super long to resolve. It means that the 9 hints will be displayed only when the last hint which takes time will be reolved.

If the inlayHints return the 10 hints it will display the 9 hint quickly and the last hint which takes time will appear when resolveInlayHint will be finished.